Fixes #2823 remove redundant NOT_INITILIZED evens and change event or…

danchom opened the pull request:
…der.

Signed-off-by: Yordan Mihaylov <j.mihailov@prosyst.bg>

kaikreuzer commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java:
```diff
@@ -504,8 +504,8 @@ public void added(Provider<Rule> provider, Rule element) {
             String rUID = ruleEngine.getUniqueId();
             ruleWithUID = initRuleId(rUID, element);
         }
-        super.added(provider, ruleWithUID);
```
Is it safe to change this order? 
You are now sending the event, before the rule is available in the registry - why is this necessary?

danchom replied:
It is not available in registry, but it is already stored in RuleManagedProvider. IMHO added/updated events should be before the state events of rule. Otherwise we can  receive a state events for rules which are not available in the registry (when the add method is failed)

kaikreuzer replied:
Looking at the code, I would actually suggest to move the "post event" line to `notifyListenersAboutXXXXElement` - this is where events are sent in ItemRegistryImpl and ThingRegistryImpl, so doing the very same here would imho make sense, wdyt?

danchom replied:
It will not be enough, because the status events are fired when the rule is added into RuleEngine, which is done in the onAddElement() method at the moment. The logic should be moved from onXXXElement() to  notifyListenersAboutXXXXElement(). I'll try to fix this.
BTW wdyt about removing redundant NOT_INITIALIZED event.
Up to now we have:
added -> not_initillized -> idle
the proposal is:
added -> idle 
or 
added -> not_initilized.

kaikreuzer replied:
> BTW wdyt about removing redundant NOT_INITIALIZED event.

You are referring to [this suggetion](https://github.com/eclipse/smarthome/issues/2823), right? If so, please see my comment on that issue.

danchom commented:
Implemented:
added -> UNINITIALIZED(without event) -> INITIALIZING -> IDLE
added -> UNINITIALIZED(without event) -> INITIALIZING -> UNINITIALIZED (with details)
updated-> UNINITIALIZED(without event) -> INITIALIZING -> IDLE
updated ->UNINITIALIZED(without event) -> INITIALIZING -> UNINITIALIZED (with details)
enabled -> DISABLED ->UNINITIALIZED(without event) -> INITIALIZING -> IDLE
enabled -> DISABLED ->UNINITIALIZED(without event) -> INITIALIZING ->  UNINITIALIZED (with details)

danchom commented:
Could you please review this PR.
